### PR TITLE
Add updated eggnog mapper tool

### PR DESCRIPTION
**Pull request by caleb-easterly**
This PR adds an updated version of the `eggnog-mapper` tool, which maps FASTA sequences to the eggNOG database, thereby transferring functional assignment by orthology. The new version uses conda to get `eggnog-mapper`, and includes test data from [egglet](https://github.com/caleb-easterly/egglet).

### CODE REVIEW

File: tools/eggnog_mapper/eggnog_mapper.xml [hunk 1]
```diff
@@ -0,0 +1,491 @@
...
+                            <param name="hmm_score" type="integer" min="0" value="20" label="Bit score threshold" />
+                            <param name="hmm_maxseqlen" type="integer" value="5000" label="Ignore query sequences larger than `maxseqlen`.  Default=5000" />
+                            <param name="hmm_qcov" type="float" value="" min="0." max="1." optional="true" label="min query coverage (from 0 to 1). Default=(disabled)" />
+                            <param name="Z" type="integer" value="40000000" min="0" label="Fixed database size used in phmmer/hmmscan (allows comparing e-values among databases).  Default=40,000,000" />
+                        </when>
+                </conditional>
+                </when>
+                <when value="diamond">
+                    <!-- db.database is used in diamond mode only to name outputs -->
+                    <param name="database" type="hidden" value="diamond"/>
```
Comment (bgruening) [inline tools/eggnog_mapper/eggnog_mapper.xml]: why is this needed?

File: tools/eggnog_mapper/eggnog_mapper.xml [hunk 2]
```diff
@@ -0,0 +1,491 @@
...
+                <option value="all" selected="true">all</option>
+            </param>
+            <param name="go_evidence" type="select"
+                    label="Select the set of GO terms that should be used for annotation">
+                <option value="experimental">experimental = Use only terms inferred from experimental evidence</option>
+                <option value="non-electronic" selected="true">non-electronic = Use only non-electronically curated terms</option>
+            </param>
+        </section>
+        <section name="seed_ortholog_options" expanded="false" title="Seed Ortholog Search Options">
+            <param name="seed_ortholog_evalue" type="float" value="0.001" label="Min E-value threshold">
```
Comment (bgruening) [inline tools/eggnog_mapper/eggnog_mapper.xml]: min=0?

File: tools/eggnog_mapper/eggnog_mapper.xml [hunk 3]
```diff
@@ -0,0 +1,491 @@
...
+                <help>
+                Min E-value expected when searching for seed eggNOG ortholog. Applies to phmmer/diamond searches. 
+                Queries not having a significant seed orthologs will not be annotated.
+                </help>
+            </param>
+            <param name="seed_ortholog_score" type="integer" value="60" label="Minimum bit score threshold">
```
Comment (bgruening) [inline tools/eggnog_mapper/eggnog_mapper.xml]: min=0

File: tools/eggnog_mapper/eggnog_mapper.xml [hunk 4]
```diff
@@ -0,0 +1,491 @@
...
+            <param name="no_annot" type="boolean" truevalue="--no_annot" falsevalue="" checked="false"
+                label="Skip functional annotation, reporting only hits"/>
+            <param name="report_orthologs" type="boolean" truevalue="--report_orthologs" falsevalue="" checked="false"
+                label="Output a file with the list of orthologs for each hits"/>
+        </section>
+    </inputs>
+    <outputs>
+        <data name="seed_orthologs" format="tabular" label="${tool.name} on ${on_string}: ${db.database}.seed_orthologs" from_work_dir="results.emapper.seed_orthologs">
+            <actions>
+                <action name="column_names" type="metadata" default="query_name,seed_eggNOG_ortholog,seed_ortholog_evalue,seed_ortholog_score"/>
```
Comment (bgruening) [inline tools/eggnog_mapper/eggnog_mapper.xml]: Oh, nice - we should make use more of this!!!

File: tools/eggnog_mapper/eggnog_mapper.xml [hunk 5]
```diff
@@ -0,0 +1,491 @@
...
+        #end if
+        $output_options.no_file_comments
+        $output_options.no_annot
+        $output_options.report_orthologs
+        --output='results'
+        -i '${input}'
+    ]]></command>
+        <inputs>
+            <!-- used for testing -->
+            <param name="test" type="hidden" value="false"/>
```
Comment (bgruening) [inline tools/eggnog_mapper/eggnog_mapper.xml]: I don't understand why this is needed?

File: tools/eggnog_mapper/eggnog_mapper.xml [hunk 6]
```diff
@@ -0,0 +1,491 @@
+<tool id="eggnog_mapper" name="eggNOG Mapper" version="0.2.0">
```
[the remark this answers is not in the document]
Comment (caleb-easterly) [reply tools/eggnog_mapper/eggnog_mapper.xml]: Do wrapper versions typically match tool versions?

[the remark this answers is not in the document]
Comment (bgruening) [reply tools/eggnog_mapper/eggnog_mapper.xml]: We try to follow this, yes. So that the user reports an error we know approx what the underlying tool version is.

File: tools/eggnog_mapper/eggnog_mapper.xml [hunk 7]
```diff
@@ -0,0 +1,491 @@
...
+    <description>functional sequence annotation by orthology</description>
+    <requirements>
+        <requirement type="package" version="1.0.1">eggnog-mapper</requirement>
+    </requirements>
+    <stdio>
```
[the remark this answers is not in the document]
Comment (caleb-easterly) [reply tools/eggnog_mapper/eggnog_mapper.xml]: is that in `<command>`? Like `<command detect_errors="exit_code">` or `<command detect_errors="aggressive"`?

[the remark this answers is not in the document]
Comment (bgruening) [reply tools/eggnog_mapper/eggnog_mapper.xml]: Yes :)

File: tools/eggnog_mapper/eggnog_mapper.xml [hunk 8]
```diff
@@ -0,0 +1,491 @@
...
+                --hmm_evalue=$db.hmm_options.hmm_evalue
+                --hmm_score=$db.hmm_options.hmm_score
+                --hmm_maxseqlen=$db.hmm_options.hmm_maxseqlen
+                #if str($db.hmm_options.hmm_qcov):
+                    --hmm_qcov=$db.hmm_options.hmm_qcov
+                #end if
+                --Z=$db.hmm_options.Z
+            #end if
+        #elif ($db.mode == 'diamond'):
+            ## #if $db.sensitivity == "1":
```
[the remark this answers is not in the document]
Comment (caleb-easterly) [reply tools/eggnog_mapper/eggnog_mapper.xml]: i forgot that was there - I took it out to avoid confusion, since sensitivity is not implemented in the `eggnog-mapper` tool.

File: tools/eggnog_mapper/eggnog_mapper.xml (hunk 4 above)
Comment (caleb-easterly) [reply tools/eggnog_mapper/eggnog_mapper.xml]: that was all @jj-umn :)

File: tools/eggnog_mapper/eggnog_mapper.xml (hunk 1 above)
Comment (caleb-easterly) [reply tools/eggnog_mapper/eggnog_mapper.xml]: this is used to automatically name the outputs in 

`${tool.name} on ${on_string}: ${db.database}.seed_orthologs`,

because diamond mode doesn't require a database name to be passed

File: tools/eggnog_mapper/eggnog_mapper.xml (hunk 5 above)
Comment (caleb-easterly) [reply tools/eggnog_mapper/eggnog_mapper.xml]: good catch! this was left over from my earlier attempts to make testing work

**Comment by caleb-easterly** [comment]
thanks for the review, @bgruening! i think I addressed everything in the latest commit.

**Comment by bgruening** [comment]
Mh, now tests are failing :(
Thanks a lot @caleb-easterly! I will wait for @jj-umn to have a final look over it.

**Comment by caleb-easterly** [comment]
Yes, it turns out I was using that `test` param for some hacky path operations. I am trying to do it in a cleaner way now, and I'll push again once I figure it out.

**Comment by caleb-easterly** [comment]
@bgruening, it's a different diamond database than would normally be used in a stand-alone diamond search (they've taken the proteins in eggNOG and diamondized them), so I don't think it would be that useful outside of the eggnog-mapper context.

**Comment by bgruening** [comment]
@caleb-easterly thanks for the explanation. If you can wait I will hope that @jj-umn will look over it before merging.

**Comment by jj-umn** [comment]
@caleb-easterly  Looks good.  @bgruening Thanks for all the help and careful reviews.
